presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java

```diff
@@ -288,25 +282,107 @@ public void setWindowFunctions(QuerySpecification node, List<FunctionCall> funct
         return windowFunctions.get(query);
     }
 
-    public void setOutputDescriptor(RelationType descriptor)
+    public Scope getScope(Node node)
     {
-        outputDescriptor = descriptor;
+        return tryGetScope(node).orElseThrow(() -> new IllegalArgumentException(String.format("Analysis does not contain information for node: %s", node)));
     }
 
-    public RelationType getOutputDescriptor()
+    public Optional<Scope> tryGetScope(Node node)
+    {
+        // TODO check performance impact for planning of getScope method
+        if (scopes.containsKey(node)) {
+            return Optional.of(scopes.get(node));
+        }
+
+        Optional<Node> root = roots.stream()
+                .filter(i -> nodeContains(i, node))
+                .findAny();
+
+        if (root.isPresent()) {
+            GetScopeVisitor visitor = new GetScopeVisitor(scopes, node);
+            visitor.process(root.get(), new LinkedList<>());
+            return visitor.getResult();
+        }
+        return Optional.empty();
+    }
+
+    public Node getRoot()
     {
-        return outputDescriptor;
+        return roots.get(0);
     }
 
-    public void setOutputDescriptor(Node node, RelationType descriptor)
+    public Scope getRootScope()
+    {
+        return getScope(getRoot());
+    }
+
+    private static class GetScopeVisitor
+            extends DefaultTraversalVisitor<Void, LinkedList<Scope>>
+    {
+        private final IdentityHashMap<Node, Scope> scopes;
+        private final Node node;
+        private Optional<Scope> result = Optional.empty();
+
+        public GetScopeVisitor(IdentityHashMap<Node, Scope> scopes, Node node)
+        {
+            this.scopes = requireNonNull(scopes, "scopes is null");
+            this.node = requireNonNull(node, "node is null");
+        }
+
+        @Override
+        public Void process(Node current, LinkedList<Scope> stack)
```

Comment:
Why do we need a stack? Wouldn't it be sufficient to just pass the "candidate" scope as the tree is walked recursively (i.e., leverage the call stack, instead)

Something like:

```java
Void process(Node current, Scope candidate)
{
    if (result.isPresent()) {
        return null;
    }

    Scope newCandidate = scopes.get(current);
    if (newCandidate == null) {
         newCandidate = candidate;
    }

    if (node == current) {
        result = Optional.of(newCandidate);
        return null;
    }

    return super.process(current, newCandidate);
}
```